design/baremetal-operator/raid-disk-controller.md

```diff
@@ -0,0 +1,193 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# raid-extension-for-physical-disks-and-controller
+
+## Status
+
+provisional
+
+## Summary
+
+implementation for the same.
+
+## Motivation
+
+The user should be able to specify exactly which physical disks and RAID
+controllers are to be used to construct the hardware RAID volume(s).
+
+### Goals
+
+The primary goal is to depict the physical disk and RAID controller names in
+the baremetal-host ``hardware-raid`` section.
+
+Afterwards, implementation of the same will be done within the
+``baremetal-operator`` by extending the ``baremetal-host`` specification.
+
+### Non-Goals
+
+- This specification does not deal with ``software-raid`` and extension for it.
+- It does not attempt to cover any generic (vendor agnostic)  naming convention
+  for disks or controllers.
+- It does not cover testing for hardware from all the vendors. It will only be
+  tested on Dell EMC hardware, and other vendors will have to test it on their
+hardware.
+
+## Proposal
+
+The proposal is to allow specifying the physical disks and RAID controller
+names in the hardware RAID section. For this purpose, new YAML attributes are
+to added in the BareMetalHost spec, namely `physicalDisks` and `controller`.
+
+In terms of the CRD, the current implementation of the RAID has the following
+sections in it, for hardware RAID:
+
+```yaml
+spec:
+  raid:
+    hardwareRAIDVolumes:
+      - sizeGibiBytes: 1000
+        level: 1
+        name: <volume-name>
+        rotational: true
+        numberOfPhysicalDisks: 2
+```
+
+We propose to add fields into the CRD, in the following manner:
+
+```yaml
+spec:
+  raid:
+    hardwareRAIDVolumes:
+      - sizeGibiBytes: 1000
+        level: 1
+        name: <volume-1-name>
+        rotational: true
+        numberOfPhysicalDisks: 2
+        physicalDisks:
+          - <disk-1-name>
+          - <disk-2-name>
```

Comment:
Just thinking aloud on the decision to make the API for `physicalDisks` a list of strings here.

In gophercloud, the `physicalDisks` is of type `[]interface{}` (https://github.com/gophercloud/gophercloud/blob/ad8066cf841fa16a5497835d6df808bf7aadcbe4/openstack/baremetal/v1/nodes/requests.go#L605-L606), not `[]string`.

That's because in Ironic the `physical_disks` parameter can either be a list of device names (like you're using here) or a list of pairs for device hints: https://github.com/openstack/ironic/blob/af0e5ee096fa237290776969a37f3bced96b7456/ironic/drivers/raid_config_schema.json#L59-L63

I think the hints are more for software RAID. I guess if there was a hardware RAID manager that used hints, we could have `physicalDiskHints` here in the API and then handle passing that to gophercloud later, so I _think_ it's all good for `physicalDisks` here to be a list of strings for the names of the disks.

Reply:
Good observation! I had to convert the list of strings to a list of interfaces while implementing this. Since the exact physical disk names needed to be used in our use case, using strings made more sense to me.